fms_extras/utils/generation.py

```diff
@@ -0,0 +1,141 @@
+
+    Args:
+        model: Callable or nn.Module
+            A function or nn.Module that takes a batch of input_ids and returns logits
+        kv_cache_manager: PagedKVCacheManager
+            the paged KVCacheManager that handles management of the kv-cache for paged attention
+        max_seq_len: int
+            the max sequence length of the model
+        max_new_tokens: int
+            max tokens to generate
+        temperature: float
+            temperature of softmax when sampling
+        top_k: int
+            only search among top k tokens
+        do_sample: bool
+            multinomial sampling. False for greedy.
+        decode_model: Callable or nn.Module, optional
+            a model to used specifically for decode step. If not given, the model input will be used for decode step
+
+    Returns:
+    Tuple[torch.Tensor, int, int]
+        the resulting output tokens, the number of new tokens generated, and the time it took per token in seconds
+    """
+    if decode_model is None:
+        decode_model = model
+
+    bsize = len(input_ids)
+
+    if cudagraphs and bsize != 1:
+        raise NotImplementedError(
+            "cudagraphs is not yet supported for batch sizes greater than 1"
+        )
+
+    # Build padded batched input tensor
+    max_len = max([seq.size(0) for seq in input_ids])
+    n_pads_init = [max_len - seq.size(0) for seq in input_ids]
+    input_ids = torch.stack(
+        [F.pad(input_ids[i], (n_pads_init[i], 0)) for i in range(bsize)]
+    )
+
+    result = input_ids
+    next_input = input_ids
+    kwargs: MutableMapping[str, Any] = dict()
+    kwargs["cache_data"] = None
+    kwargs["use_cache"] = True
+    sequence_ids: Optional[List[int]] = None
+    block_mapping_max = ((max_len + max_new_tokens) // 16) + 1
+    for i in range(max_new_tokens):
+        input_ids = next_input[:, -max_seq_len:]
+
+        # compute the mask
+        if i == 0:
+            is_pad = input_ids == 0
+            mask = is_pad.unsqueeze(-1) == is_pad.unsqueeze(-2)
+            kwargs["mask"] = mask.tril(diagonal=0)
+        else:
+            if i == 1:
+                start_time = time.time()
+            kwargs["mask"] = None
+
+        # get the cache data and position ids if using cache
+        if sequence_ids is None:
+            num_tokens_per_sequence = torch.count_nonzero(input_ids.T, dim=0).tolist()
+        else:
+            num_tokens_per_sequence = [1 for _ in range(input_ids.size(0))]
+
+        cache_data = kv_cache_manager.allocate_tokens(
+            num_tokens_per_sequence, sequence_ids
+        )
+
+        sequence_ids = cache_data.sequence_ids
+
+        kwargs["cache_data"] = cache_data
+        kwargs["position_ids"] = cache_data.compute_position_ids(
+            num_tokens_per_sequence
+        )
+
+        if i == 0:
+            logits, _ = model(input_ids, **kwargs)
+        else:
+            # cudagraph requires static shapes
+            if cudagraphs:
```

Comment:
can any of the larger nested chunks of code be extracted out into helper functions? Overall this function is a bit long with a lot of nesting so I think we can find some parts to refactor out to help with readability.

Reply:
added a function for this to pad to right for cudagraphs in latest commit

Reply:
is there any more that can be refactored out?

Reply:
cleaned the generate up a bit and removed the compute_position_ids. The preprocessing for model_kwargs now should be pretty minimal